day2/src/main.rs

```diff
@@ -0,0 +1,59 @@
+use itertools::Itertools;
+use std::time::Instant;
+
+const INPUT: &[u8] = include_bytes!("input.txt");
+
+fn parse(array: &[u8]) -> i32 {
+    if array.len() == 1 {
+        return (array[0] - b'0') as i32;
+    }
+    return (array[0] - b'0') as i32 * 10 + (array[1] - b'0') as i32;
+}
+
+fn report_is_safe(report: &Vec<i32>) -> bool {
+    let asc = report[1] > report[0];
+    report
+        .windows(2)
+        .find(|w| {
+            let diff = if asc { w[1] - w[0] } else { w[0] - w[1] };
+            diff < 1 || diff > 3
+        })
+        .is_none()
+}
+
+fn part1() -> usize {
+    INPUT
+        .split(|&b| b == b'\n')
+        .map(|report| report.split(|&b| b == b' ').map(parse).collect())
```

Comment:
If you really wanted to go to the extreme here you could possibly save some allocations by collecting into a special data structure which has a max length of the max number of elements per line rather than a vector which has to be heap allocated.

```rust
struct LineNumbers {
    len: usize,
    data: [i32; MAX_NUMBERS_PER_LINE]
}
```

You would then have an [as_ref](https://doc.rust-lang.org/std/convert/trait.AsRef.html#tymethod.as_ref)(&self) -> &[[T]](https://doc.rust-lang.org/std/primitive.slice.html) function that returns a slice of the correct length. Since you've already changed `report_is_safe` to accept a slice the rest of the function should continue to work.